server/tests/e2e/decentralized.spec.ts

```diff
@@ -114,4 +119,148 @@ describe('end-to-end decentralized', function () {
+
+    // Have User 2 join the task and train for one round
+    const logUser2Round1 = await generatorUser2.next()
+    expect(logUser2Round1.done).to.be.false
+    // participant list not updated yet (updated at step c))
+    expect((logUser2Round1.value as RoundLogs).participants).equal(1)
+    // User 2 did a) and b)
+    expect(statusUser2.last()).equal("TRAINING")
+    // User 1 is still in c) now waiting for user 2 to be ready to exchange weight updates
+    expect(statusUser1.last()).equal("RETRIEVING PEERS")
+
+    // Proceed with round 2
+    // The server should answer with the round's peers list. 
+    // Peers then exchange updates and then start training locally with the new weights
+    const logUser2Round2 = await generatorUser2.next()
+    const logUser1Round2 = await logUser1Round2Promise // the promise can resolve now
+    expect(logUser1Round2.done).to.be.false
+    expect(logUser2Round2.done).to.be.false
+    // nb of participants should now be updated
+    expect((logUser1Round2.value as RoundLogs).participants).equal(2)
+    expect((logUser2Round2.value as RoundLogs).participants).equal(2)
+    // User 1 and 2 did c), a) and b)
+    expect(statusUser1.pop().last()).equal("UPDATING MODEL") // second to last
+    expect(statusUser1.last()).equal("TRAINING")
+    expect(statusUser2.pop().last()).equal("UPDATING MODEL")
+    expect(statusUser2.last()).equal("TRAINING")
+    
+    // Have user 1 quit the session
+    await discoUser1.close()
+    // Make user 2 go to c)
+    const logUser2Round3Promise = generatorUser2.next()
+    await new Promise((res, _) => setTimeout(res, statusUpdateTime)) // Wait some time for the status to update
+    expect(statusUser2.last()).equal("NOT ENOUGH PARTICIPANTS")
+
+    // Create User 3
+    const discoUser3 = new Disco(lusCovidTask, url, { scheme: "decentralized" });
+    let statusUser3 = List<RoundStatus>();
+    discoUser3.on("status", status => { statusUser3 = statusUser3.push(status) })
+    const generatorUser3 = discoUser3.trainByRound(["image", dataset])
+
+    // User 3 joins mid-training and trains one local round
+    const logUser3Round1 = await generatorUser3.next()
+    expect(logUser3Round1.done).to.be.false
+    // participant list not updated yet
+    expect((logUser3Round1.value as RoundLogs).participants).equal(1)
+    // User 3 did a) and b)
+    expect(statusUser3.last()).equal("TRAINING")
+    // User 2 is still in c) waiting for user 3 to be ready to exchange waits
+    expect(statusUser2.last()).equal("RETRIEVING PEERS")
+    
+    // User 3 notifies the server that they are ready to exchange waits
+    // then user 2 and 3 exchange weight updates
+    const logUser3Round3 = await generatorUser3.next()
+    const logUser2Round3 = await logUser2Round3Promise // the promise can resolve now
+    if (logUser3Round3.done || logUser2Round3.done)
+      throw Error("User 1 or 2 finished training at the 3nd round")
```

Comment:
yeah, that's the way to check for iteration results state. TypeScript doesn't infer that `iff not done then value is valid` with function, only via `if` (it would require way more sophisticated typing means).
that allow to not cast to `RoundLogs` (as done above)